### PR TITLE
Pagination on candidates table

### PR DESCRIPTION
**Pull request by domokoslaci94**
[Trello Card](https://trello.com/c/Lh6jFfAf)

### CODE REVIEW

File: service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java [hunk 1]
```diff
@@ -83,6 +87,51 @@ public CandidateDTO getCandidateByApplicationID(Long applicationID) {
...
+        converterService.convert(candidateEntitiesPage.getContent(), CandidateDTO.class);
+
+    PagingResponse<CandidateDTO>
+        result =
+        new PagingResponse<>(candidateEntitiesPage.getTotalElements(), candidateDTOs);
+
+    return result;
+  }
+
+
```
Comment (lamfalusy) [inline service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java]: Delete this redundant and unnecessary empty line.

File: service/src/main/java/com/epam/rft/atsy/service/CandidateService.java [hunk 2]
```diff
@@ -1,7 +1,9 @@
 package com.epam.rft.atsy.service;
```
Comment (lamfalusy) [inline service/src/main/java/com/epam/rft/atsy/service/CandidateService.java:1]: As the commit description:  "Introduced new method in persistance layer to support pagination", but you introduce new method in service layer. Anyway it is not a big problem now, and you have to do nothing, but please be more consistent next time.

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 3]
```diff
@@ -280,6 +297,137 @@ public void getAllCandidateShouldCallRepositoryWithDescendingSortWhenPassingDesc
   }
 
   @Test(expected = IllegalArgumentException.class)
+  public void getCandidatesByFilterRequestShouldThrowIllegalARgumentExceptionWhenCandidateFilterRequestIsnull() {
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Typo: ...ThrowIllegalARgumentException...

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 4]
```diff
@@ -280,6 +297,137 @@ public void getAllCandidateShouldCallRepositoryWithDescendingSortWhenPassingDesc
...
+    // Given
+
+    // When
+    candidateService.getCandidatesByFilterRequest(null);
+    // Then
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void getCandidatesByFilterRequestShouldThrowIllegalARgumentExceptionWhenPageSizeIsnull() {
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Typo: ...ThrowIllegalARgumentException...

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 5]
```diff
@@ -280,6 +297,137 @@ public void getAllCandidateShouldCallRepositoryWithDescendingSortWhenPassingDesc
...
+    CandidateFilterRequest
+        candidateFilterRequest =
+        CandidateFilterRequest.builder().pageSize(null).build();
+    // When
+    candidateService.getCandidatesByFilterRequest(candidateFilterRequest);
+    // Then
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void getCandidatesByFilterRequestShouldThrowIllegalARgumentExceptionWhenPageNumberIsnull() {
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Typo: ...ThrowIllegalARgumentException...

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 6]
```diff
@@ -280,6 +297,137 @@ public void getAllCandidateShouldCallRepositoryWithDescendingSortWhenPassingDesc
...
+    CandidateFilterRequest
+        candidateFilterRequest =
+        CandidateFilterRequest.builder().pageNumber(null).build();
+    // When
+    candidateService.getCandidatesByFilterRequest(candidateFilterRequest);
+    // Then
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void getCandidatesByFilterRequestShouldThrowIllegalARgumentExceptionWhenSortNameIsNotNullButSortOrderIsNull() {
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Typo: ...ThrowIllegalARgumentException...

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 7]
```diff
@@ -280,6 +297,137 @@ public void getAllCandidateShouldCallRepositoryWithDescendingSortWhenPassingDesc
...
+    CandidateFilterRequest
+        candidateFilterRequest =
+        CandidateFilterRequest.builder().sortName(SORT_NAME_NAME).sortOrder(null).build();
+    // When
+    candidateService.getCandidatesByFilterRequest(candidateFilterRequest);
+    // Then
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void getCandidatesByFilterRequestShouldThrowIllegalARgumentExceptionWhenSortOrderIsNotNullButSortNameIsNull() {
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Typo: ...ThrowIllegalARgumentException...

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 8]
```diff
@@ -280,6 +297,137 @@ public void getAllCandidateShouldCallRepositoryWithDescendingSortWhenPassingDesc
...
+    CandidateFilterRequest
+        candidateFilterRequest =
+        CandidateFilterRequest.builder().sortOrder(SORT_ORDER_ASC).sortName(null).build();
+    // When
+    candidateService.getCandidatesByFilterRequest(candidateFilterRequest);
+    // Then
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void getCandidatesByFilterRequestShouldThrowIllegalARgumentExceptionWhenSortOrderIsInvalid() {
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Typo: ...ThrowIllegalARgumentException...

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 9]
```diff
@@ -280,6 +297,137 @@ public void getAllCandidateShouldCallRepositoryWithDescendingSortWhenPassingDesc
...
+        candidateFilterRequest =
+        CandidateFilterRequest.builder().sortName(SORT_NAME_NAME).sortOrder(INVALID_SORT_ORDER)
+            .build();
+    // When
+    candidateService.getCandidatesByFilterRequest(candidateFilterRequest);
+    // Then
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void getCandidatesByFilterRequestShouldThrowIllegalARgumentExceptionWhenSortNameIsInvalid() {
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Typo: ...ThrowIllegalARgumentException...

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 10]
```diff
@@ -280,6 +297,137 @@ public void getAllCandidateShouldCallRepositoryWithDescendingSortWhenPassingDesc
...
+        pageRequest =
+        new PageRequest(candidateFilterRequest.getPageNumber(),
+            candidateFilterRequest.getPageSize(),
+            Sort.Direction.fromString(candidateFilterRequest.getSortOrder()),
+            candidateFilterRequest.getSortName());
+
+    String name = candidateFilterRequest.getCandidateName();
+    String email = candidateFilterRequest.getCandidateEmail();
+    String phone = candidateFilterRequest.getCandidatePhone();
+    String positions = candidateFilterRequest.getCandiadtePositions();
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Minor improvement suggestion: these local variables (pageRequest, name, email, phone, positions) can be final.

File: service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java [hunk 11]
```diff
@@ -114,11 +117,23 @@ public CandidateDTO getCandidateByApplicationID(Long applicationID) {
...
+        MoreObjects.firstNonNull(candidateFilterRequest.getCandidateName(), StringUtils.EMPTY);
+    String
+        email =
+        MoreObjects.firstNonNull(candidateFilterRequest.getCandidateEmail(), StringUtils.EMPTY);
+    String
+        phone =
+        MoreObjects.firstNonNull(candidateFilterRequest.getCandidatePhone(), StringUtils.EMPTY);
+    String
+        positions =
+        MoreObjects.firstNonNull(candidateFilterRequest.getCandiadtePositions(), StringUtils.EMPTY);
```
Comment (lamfalusy) [inline service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java:109]: Minor improvement suggestion: local variables (name, email, phone, position, candidateEntitiesPage, candidateDTOs, result) can be final.

File: web/src/main/java/com/epam/rft/atsy/web/controllers/rest/CandidateController.java [hunk 12]
```diff
@@ -28,65 +27,56 @@
...
-    FilterRequest filterRequest = FilterRequest.builder()
-        .order(orderDirection)
-        .fieldName(fieldName)
-        .searchOptions(parseFilters(filter))
-        .build();
+    SearchOptions searchOptions = parseFilters(filter);
+    String name = MoreObjects.firstNonNull(searchOptions.getName(), StringUtils.EMPTY);
+    String email = MoreObjects.firstNonNull(searchOptions.getEmail(), StringUtils.EMPTY);
+    String phone = MoreObjects.firstNonNull(searchOptions.getPhone(), StringUtils.EMPTY);
+    String position = MoreObjects.firstNonNull(searchOptions.getPosition(), StringUtils.EMPTY);
```
Comment (lamfalusy) [inline web/src/main/java/com/epam/rft/atsy/web/controllers/rest/CandidateController.java:65]: Minor improvement suggestion: local variables (name, email, phone, position, candidateFilterRequest) can be final.

File: service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java [hunk 13]
```diff
@@ -416,17 +285,13 @@ public void getCandidatesByFilterRequestShouldFindDummyCandidateDTOInAscendingOr
         actualPagingResponse =
         candidateService.getCandidatesByFilterRequest(candidateFilterRequest);
 
-    // Then
```
Comment (lamfalusy) [inline service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java]: Why did you delete this comment?

File: service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java [hunk 14]
```diff
@@ -107,9 +107,9 @@ public CandidateDTO getCandidateByApplicationID(Long applicationID) {
     String
         positions =
         MoreObjects.firstNonNull(candidateFilterRequest.getCandiadtePositions(), StringUtils.EMPTY);
+    Page<CandidateEntity> candidateEntitiesPage;
 
-    Page<CandidateEntity>
-        candidateEntitiesPage =
+    candidateEntitiesPage =
```
Comment (lamfalusy) [inline service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java]: Why did you this modification? For me it seems unnecessary.

File: service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java [hunk 15]
```diff
@@ -183,10 +182,18 @@ private void validateCandidateFilterRequest(CandidateFilterRequest candidateFilt
 
       try {
         Sort.Direction.fromString(candidateFilterRequest.getSortOrder());
-      } catch (IllegalArgumentException e) {
+      } catch (Exception e) {
         throw new IllegalArgumentException(
             "Invalid sort order: " + candidateFilterRequest.getSortOrder(), e);
       }
+
+      try {
```
Comment (lamfalusy) [inline service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java:167]: I do not really understand the reason of this try block. Is it necessary? A new PageRequest object is created but nothing to do with it.

File: persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/CandidateRepositoryIT.java [hunk 16]
```diff
@@ -336,4 +337,26 @@ public void findByEmailShouldFindExistingCandidate() {
...
     assertThat(candidate.getEmail(), is(EXISTING_CANDIDATE_EMAIL));
   }
+
+  @Test
+  public void findByCandidateFiletRequestShouldFindCandidateAOnlyByPhoneFilter() {
+    //Given
+    String name = "";
+    String email = "";
+    String phone = CANDIDATE_A_PHONE;
+    String position = "";
```
Comment (lamfalusy) [inline persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/CandidateRepositoryIT.java]: Minor improvement suggestion: these local variables (name, email, phone, position, expectedCandidates, actualCandidates) can be final.

File: persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/CandidateRepositoryIT.java [hunk 17]
```diff
@@ -126,6 +138,47 @@ private void assertCandidateList(List<CandidateEntity> expectedCandidates,
...
+    assertCandidateList(expectedCandidates, actualCandidate);
+  }
+
+  @Test
+  public void findByCandidateFieldRequestShouldFindCandidateCOnlyByPositionFilter() {
+    //Given
+    String name = "";
+    String email = "";
+    String phone = "";
+    String position = "f";
```
Comment (lamfalusy) [inline persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/CandidateRepositoryIT.java]: Please refactor "f" string literal to a constant. Could you please do it with other string literals in this test class?

File: service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java (hunk 15 above)
Comment (domokoslaci94) [reply service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java:167]: This block handles the validation of the page request, e.g.: if the sortName  argument is malformed an exception is thrown.

Comment (lamfalusy) [reply service/src/main/java/com/epam/rft/atsy/service/impl/CandidateServiceImpl.java:167]: All right, i got it.
